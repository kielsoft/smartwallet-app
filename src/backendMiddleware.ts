import { JolocomLib } from 'jolocom-lib'
import { IdentityWallet } from 'jolocom-lib/js/identityWallet/identityWallet'
import { EthereumLib, EthereumLibInterface } from 'src/lib/ethereum'
import { EncryptionLib, EncryptionLibInterface } from 'src/lib/crypto'
import { Storage } from 'src/lib/storage/storage'
import { KeyChain, KeyChainInterface } from 'src/lib/keychain'
import { ConnectionOptions } from 'typeorm/browser'

// TODO Type config better
interface BackendMiddlewareConfig {
    fuelingEndpoint: string, 
    typeOrmConfig: ConnectionOptions
}

export class BackendMiddleware {
  identityWallet!: IdentityWallet
  ethereumLib: EthereumLibInterface
  storageLib: Storage
  encryptionLib: EncryptionLibInterface
  keyChainLib: KeyChainInterface

  constructor(config: BackendMiddlewareConfig) {
    this.ethereumLib = new EthereumLib(config.fuelingEndpoint)
    this.storageLib = new Storage(config.typeOrmConfig),
    this.encryptionLib = new EncryptionLib(),
    this.keyChainLib = new KeyChain()
  }

  async setIdentityWallet(privKey: Buffer): Promise<void> {
    const registry = JolocomLib.registry.jolocom.create()
    this.identityWallet = await registry.authenticate(privKey)
  }
}