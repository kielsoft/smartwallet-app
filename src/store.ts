import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import { entityList } from 'src/lib/storage/entities'
import { defaultConfig } from 'jolocom-lib/js/defaultConfig'
import { RootState, rootReducer } from 'src/reducers'
import { BackendMiddleware } from 'src/backendMiddleware'
const { createReactNavigationReduxMiddleware } = require('react-navigation-redux-helpers')
import { ConnectionOptions } from 'typeorm/browser'

// TODO Break apart
const typeOrmConfig:ConnectionOptions = {
    type: 'react-native',
    database: 'LocalSmartWalletData',
    location: 'default',
    logging: ['error', 'query', 'schema'],
    synchronize: true,
    entities: entityList
  }


const config = {
  defaultConfig,
  fuelingEndpoint: 'https://faucet.jolocom.com/request',
  typeOrmConfig
}

createReactNavigationReduxMiddleware('root', (state : RootState) => state.navigation)
const backendMiddleware = new BackendMiddleware(config)

export const store = createStore(
  rootReducer,
  {},
  applyMiddleware(thunk.withExtraArgument(backendMiddleware))
)
