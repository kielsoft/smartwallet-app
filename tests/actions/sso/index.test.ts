import { JolocomLib } from 'jolocom-lib'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { CredentialRequestPayload } from 'jolocom-lib/js/interactionFlows/credentialRequest/credentialRequestPayload';

import { ssoActions } from '../../../src/actions'
import { StateCredentialRequestSummary } from '../../../src/reducers/sso/'

describe('SSO action creators', () => {
    const initialState = {
        account: {
            claims: {
                toJS: () => {
                    return {
                        loading: false,
                        selected: {
                            credentialType: 'Email',
                            claimData: {
                                email: 'kielsoft@gmail.com'
                            },
                            id: '',
                            issuer: 'did:jolo:test',
                            subject: 'did:jolo:test'
                        },
                        pendingExternal: [],
                        decoratedCredentials: 'blah'
                    }
                }
            },
            did: {
                toJS: () => ({ did: 'mock:did:test' })
            }
        },
        activeCredentialRequest: {
            requester: '',
            callbackURL: '',
            availableCredentials: []
        }
    }

    const mockStore = configureStore([thunk])(initialState)
    const encodedJwt = "eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJjcmVkZW50aWFsUmVxdWVzdCI6eyJjcmVkZW50aWFsUmVxdWlyZW1lbnRzIjpbeyJ0eXBlIjpbIkNyZWRlbnRpYWwiLCJQcm9vZk9mRW1haWxDcmVkZW50aWFsIl0sImNvbnN0cmFpbnRzIjp7ImFuZCI6W3siPT0iOlt0cnVlLHRydWVdfSx7Ij09IjpbdHJ1ZSx0cnVlXX1dfX0seyJ0eXBlIjpbIkNyZWRlbnRpYWwiLCJQcm9vZk9mTmFtZUNyZWRlbnRpYWwiXSwiY29uc3RyYWludHMiOnsiYW5kIjpbeyI9PSI6W3RydWUsdHJ1ZV19LHsiPT0iOlt0cnVlLHRydWVdfV19fV0sImNhbGxiYWNrVVJMIjoiaHR0cDovL2xvY2FsaG9zdDo5MDAwL2F1dGhlbnRpY2F0aW9uLzJhM2hnIn0sInR5cCI6ImNyZWRlbnRpYWxSZXF1ZXN0IiwiaWF0IjoxNTQxODY3OTE5MTI5LCJpc3MiOiJkaWQ6am9sbzpiMzEwZDI5M2FlYWM4YTVjYTY4MDIzMmI5NjkwMWZlODU5ODhmZGUyODYwYTFhNWRiNjliNDk3NjI5MjNjYzg4I2tleXMtMSJ9.BIRzU5gHEmEKOEjMvHoDVI4vr7JtK8dw-gYBrvKQThzAibdsdfAuGXOtkzsBp30o2A4HbfPrABf1JwEEOa0zVA"
    let returnedDecodedJwt: CredentialRequestPayload = null;

    beforeEach(async () => {
        mockStore.clearActions()
        returnedDecodedJwt = await JolocomLib.parse.interactionJSONWebToken.decode(encodedJwt)
    })

    it('should set credential request json data', () => {

        const reqSummary: StateCredentialRequestSummary = {
            requester: 'did:jolo:test',
            availableCredentials: [],
            callbackURL: 'http://192.168.8.102:9000/authentication/',
        }

        const action = ssoActions.setCredentialRequest(reqSummary)
        expect(action).toMatchSnapshot();

    }, 30000)

    it('should correctly parse encoded JWT string without any Exception and consumeRedentialRequest', async () => {

        const backendMiddleware = {
            storageLib: {
                get: {
                    persona: jest.fn().mockResolvedValue([{ did: 'mock:did:test' }]),
                    attributesByType: (requestType: string[])=> {
                        switch(requestType[1]){
                            case "ProofOfEmailCredential":
                                return {"type":["Credential","ProofOfEmailCredential"],"results":[{"verification":"claimId:dcc4ced746f83","values":["kielsoft@gmail.com"],"fieldName":"email"}]};
                            
                            case "ProofOfNameCredential":
                                return {"type":["Credential","ProofOfNameCredential"],"results":[{"verification":"claimId:262cda78bfe26","values":["Olayode","Ezekiel"],"fieldName":"givenName"}]};
                            
                        }
                        return {"results":[]}
                    },
                    verifiableCredential: jest.fn().mockResolvedValue([])
                }
            }
        }

        await ssoActions.consumeCredentialRequest(returnedDecodedJwt)(mockStore.dispatch, mockStore.getState, backendMiddleware)
        expect(mockStore.getActions()).toMatchSnapshot()

    }, 30000)

})