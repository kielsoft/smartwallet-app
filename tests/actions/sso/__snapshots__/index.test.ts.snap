// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SSO action creators should correctly parse encoded JWT string without any Exception and consumeRedentialRequest 1`] = `
Array [
  Object {
    "type": "SET_CREDENTIAL_REQUEST",
    "value": Object {
      "availableCredentials": Array [
        Object {
          "type": "Email",
          "values": Array [
            "kielsoft@gmail.com",
          ],
          "verifications": Array [],
        },
        Object {
          "type": "Name",
          "values": Array [
            "Olayode",
            "Ezekiel",
          ],
          "verifications": Array [],
        },
      ],
      "callbackURL": "http://localhost:9000/authentication/2a3hg",
      "requester": "did:jolo:b310d293aeac8a5ca680232b96901fe85988fde2860a1a5db69b49762923cc88#keys-1",
    },
  },
  Object {
    "loading": false,
    "type": "SET_LOADING",
  },
  Object {
    "routeName": "Consent",
    "type": "Navigation/NAVIGATE",
  },
]
`;

exports[`SSO action creators should set credential request json data 1`] = `
Object {
  "type": "SET_CREDENTIAL_REQUEST",
  "value": Object {
    "availableCredentials": Array [],
    "callbackURL": "http://192.168.8.102:9000/authentication/",
    "requester": "did:jolo:test",
  },
}
`;
